.github/workflows/pull-request.yml

```diff
@@ -0,0 +1,152 @@
+on:
+  pull_request:
+    branches: [master]
+
+jobs:
+  build-jar:
+    runs-on: ubuntu-latest
+
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3.0.2
+
+      - name: Version
+        id: version
+        # Get version from project.clj
```

Comment:
There should be more graceful way to obtain current version

Reply:
I'm tied up today, let me get back to you tomorrow.  Thanks for keeping going on this!

Reply:
Given what I've seen in the example you provided before, I expect that this will work.  I don't have nearly the experience to look at what you've done and have any idea if it is actually correct, since this is the first workflow I've ever examined.  But that is ok.  I would like to accept this PR, but I do have one question.

It looks like it operates whenever someone does a pull request on master?

So, how do you expect this to be used?  I rarely do pull requests on master (or anything).  I expect that is in part because I'm not working with a bunch of other people who add code to the same GitHub repo.   Which could change, of course. 

The workflow you had in #249 was on push or pull with master.  I understand the "push" part, but even there I am kind of not understanding the "pull" trigger.

So I guess I'm asking what is the human workflow you imagine this pull request to be fitting into?  Could you lay out the steps you imagine someone taking to initiate this workflow after I accept this PR?

Once we work this out, I'm ready to accept this PR.

Thanks!

Reply:
This workflow is for people like me who may want to occasionally contribute to the project by creating Pull Requests. Once all checks are green, both you and me are pretty confident that new code is safe to merge.

This workflow is lightweight enough to be finished under 3–5 minutes, so if the feature is not that trivial, you can push and work further. Distraction will occur only if something went wrong.

For example, GraalVM builds may break because of missing type-hint. While we tend to check our code with unit tests before any code push, errors related to native-image builds are most likely to be discovered during release.

To sum it up, the Pull Request workflow is meant to provide quick enough feedback on the ~95% issues that may happen after particular code change.

To the another hand, workflow for the main branch push should perform more comprehensive set of checks. It will include everything included in Pull Request workflow, but also will run test for ClojureScript version that lasts 7–10 minutes. Once tests of this workflow are done, it should mean we can tag and release this commit.

Finally, release workflow will run `native-image` build on MacOS, Windows and Linux machines, so that you don't need to perform it on your own machines.

Reply:
Thanks, that is clearer.  I think my problem was that I was thinking about "push" and "pull" commands to git on my laptop relating to the master branch on GitHub.  I was wondering why I would want a workflow to run when I typed `git pull origin master` on my laptop.  But I don't believe that is what this workflow is for. 

I wasn't thinking about "pull requests", where someone wants to contribute to zprint by doing a PR.  Just to be clear (to me, at least), I think you are saying that this workflow will run when someone creates a PR for zprint?  So, in this case, it is for the person creating the PR to see if the basic tests run, and for me to also see that the PR meets the basic tests? 

Thanks.

Reply:
You're right, this meant to be the very first criteria for successful Pull Request. This is not necessary should be a human, but also can be Renovate Bot with its dependency upgrades.

One reason why people don't upgrade dependencies often IMO is that we're subconsciously afraid of unexpected work of researching which one new package update breaks our project. And if we don't have any CI, sometimes we risk to learn about breakage from the users opening an issue, which makes us feel guilty so at some point we decided: "I'll never ever will upgrade dependencies unless it's absolutely necessary".

But once this is automated, you'll transparently see which upgrade breaks the project and which is okay to merge.

By the way, I noticed from personal experience that once these PR workflows are implemented, I tend to work using Pull Requests, despite the fact direct pushes are quicker and seemed to be pragmatic for the projects with one contributor.